File: backend/app/models/algorithm_definition.py
"""Typed contracts for immutable algorithm definitions and their versions."""

from __future__ import annotations

import re
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator


SEMVER_PATTERN = re.compile(r"^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-[0-9A-Za-z.-]+)?$")
DefinitionStatus = Literal["draft", "testing", "validated_engineering", "research_use", "deprecated", "archived"]
VersionState = Literal["draft", "published"]


class DefinitionVersionDraft(BaseModel):
    semver: str
    graph: dict[str, Any]
    parameter_schema: dict[str, Any] = Field(default_factory=dict)
    inputs: dict[str, Any] = Field(default_factory=dict)
    outputs: dict[str, Any] = Field(default_factory=dict)
    units: dict[str, Any] = Field(default_factory=dict)
    quality_rules: dict[str, Any] = Field(default_factory=dict)
    references: list[str] = Field(default_factory=list)

    @field_validator("semver")
    @classmethod
    def validate_semver(cls, value: str) -> str:
        if not SEMVER_PATTERN.fullmatch(value):
            raise ValueError("semver must be MAJOR.MINOR.PATCH")
        return value


class DefinitionCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=160)
    owner: str = Field(default="local-user", min_length=1, max_length=160)
    description: str = Field(default="", max_length=4000)


class AlgorithmDefinition(BaseModel):
    definition_id: str
    name: str
    owner: str
    status: DefinitionStatus
    description: str
    created_at: str
    updated_at: str


class AlgorithmDefinitionVersion(DefinitionVersionDraft):
    version_id: str
    definition_id: str
    state: VersionState
    digest_sha256: str
    created_at: str
    published_at: str | None = None

File: frontend/src/types/algorithmDefinition.ts
export type Unit = 'V' | 'uV' | 'V^2' | 'uV^2' | 'V^2/Hz' | 'uV^2/Hz' | 'Hz' | 's' | 'ratio' | 'percent' | 'dimensionless' | 'dB re 1 uV^2/Hz'

export interface DefinitionDraft {
  semver: string
  graph: { nodes: DefinitionNode[]; outputs: string[] }
  parameter_schema: Record<string, unknown>
  inputs: Record<string, unknown>
  outputs: Record<string, unknown>
  units: Record<string, unknown>
  quality_rules: Record<string, unknown>
  references: string[]
}

export interface DefinitionNode {
  id: string
  type: string
  inputs: Record<string, string>
  parameters: Record<string, unknown>
}

export interface AlgorithmDefinition {
  definition_id: string
  name: string
  owner: string
  status: 'draft' | 'testing' | 'validated_engineering' | 'research_use' | 'deprecated' | 'archived'
  description: string
  created_at: string
  updated_at: string
}

export interface AlgorithmDefinitionVersion extends DefinitionDraft {
  version_id: string
  definition_id: string
  state: 'draft' | 'published'
  digest_sha256: string
  created_at: string
  published_at: string | null
}

export interface DefinitionCapabilities {
  nodes: string[]
  units: Unit[]
  official_execution: Record<string, 'generic_research_primitives' | 'official_composite_shadow_only'>
}

export interface PreviewRun {
  run_id: string
  status: 'completed' | 'gate_failed' | 'failed'
  is_preview: true
  requested_range: { start_s: number; end_s: number }
  actual_range: { start_s: number; end_s: number } | null
  result_summary: {
    preview: true
    outputs: Record<string, { value: number | null; unit: Unit; quality: { status: 'clean' | 'bad'; reasons: string[] } }>
  } | null
  error: { code: string; message: string } | null
}

export const DEFAULT_DRAFT: DefinitionDraft = {
  semver: '0.1.0',
  graph: { nodes: [{ id: 'out', type: 'output', inputs: { source: '$input.value' }, parameters: {} }], outputs: ['out'] },
  parameter_schema: { type: 'object', additionalProperties: false },
  inputs: { value: { type: 'Scalar', unit: 'ratio' } },
  outputs: { out: { type: 'Scalar', unit: 'ratio' } },
  units: { input: 'ratio', output: 'ratio' },
  quality_rules: {},
  references: [],
}
